compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,526 @@
+            self.inline_block(then, no_predicates);
+            let to_process = self.handle_terminator(then, &queue);
+
+            for incoming_block in to_process {
+                if !queue.contains(&incoming_block) {
+                    queue.push(incoming_block);
+                }
+            }
+        }
+
+        //2. process 'else' branch, in case there is no 'then'
+        let next = queue.pop();
+        if next == conditional.block_else {
+            let next = next.unwrap();
+            self.inline_block(next, no_predicates);
+            let _ = self.handle_terminator(next, &queue);
+        } else {
+            assert_eq!(next, Some(conditional.block_exit));
+        }
+
+        //3. process 'exit' block
+        self.inline_block(conditional.block_exit, no_predicates);
+        // Manually set the terminator of the entry block to the one of the exit block
+        let terminator =
+            self.inserter.function.dfg[conditional.block_exit].terminator().unwrap().clone();
+        let new_terminator = match terminator {
+            TerminatorInstruction::JmpIf {
+                condition,
+                then_destination,
+                else_destination,
+                call_stack,
+            } => {
+                let condition = self.inserter.resolve(condition);
+                TerminatorInstruction::JmpIf {
+                    condition,
+                    then_destination,
+                    else_destination,
+                    call_stack,
+                }
+            }
+            TerminatorInstruction::Jmp { destination, arguments, call_stack } => {
+                let arguments = vecmap(arguments, |value| self.inserter.resolve(value));
+                TerminatorInstruction::Jmp { destination, arguments, call_stack }
+            }
+            TerminatorInstruction::Return { return_values, call_stack } => {
+                let return_values = vecmap(return_values, |value| self.inserter.resolve(value));
+                TerminatorInstruction::Return { return_values, call_stack }
+            }
+        };
+        self.inserter.function.dfg.set_block_terminator(conditional.block_entry, new_terminator);
+        self.inserter.map_data_bus_in_place();
+        //4. restore the context, in case it is re-used.
+        self.target_block = old_target;
+        self.no_predicate = old_no_predicate;
+    }
+
+    fn map_block_with_mapping(
+        mapping: HashMap<ValueId, ValueId>,
+        func: &mut Function,
+        block: BasicBlockId,
+    ) {
+        // Map all instructions in the block
+        let mut inserter = FunctionInserter::new(func);
+        inserter.set_mapping(mapping);
+        let instructions = inserter.function.dfg[block].instructions().to_vec();
+        for instruction in instructions {
+            inserter.map_instruction_in_place(instruction);
+        }
+        inserter.map_terminator_in_place(block);
+    }
+
+    fn flatten_multiple(
```

Comment:
nit: I think this would be cleaner outside of the `Context` impl, but I don't feel super strongly about this.